File: python/ssrf.py
import urllib.request
from urllib.parse import urlparse
from flask import Flask, request

app = Flask(__name__)

@app.route('/fetch')
def fetch():
    url = request.args.get('url')
    if not url:
        return 'URL parameter is required', 400
    parsed = urlparse(url)
    if parsed.scheme not in ['http', 'https']:
        return 'Only HTTP and HTTPS URLs are allowed', 400
    # CWE-918: user-controlled URL fetched without validation allows SSRF.
    response = urllib.request.urlopen(url)
    return response.read()

if __name__ == '__main__':
    app.run()

File: javascript/ssrf.js
const express = require('express');
const http = require('http');
const app = express();

app.get('/fetch', (req, res) => {
  const url = req.query.url;
  // CWE-918: user-controlled URL fetched without validation allows SSRF.
  http.get(url, (response) => {
    let data = '';
    response.on('data', chunk => data += chunk);
    response.on('end', () => res.send(data));
  });
});

app.listen(3000);
